File: tests/wasm/test_smooth_theil_sen.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadTestData, assertArrayClose } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function weights(size) {
    const center = (size - 1) * 0.5;
    const width = (size - 0.5) * Math.pow(1024, -0.5) + 0.5;
    const values = Array.from({ length: size }, (_, i) =>
        Math.exp(-Math.pow((i - center) / width, 2) / 4) / Math.sqrt(2 * Math.PI),
    );
    const sum = values.reduce((a, b) => a + b, 0);
    return values.map((v) => v / sum);
}

function estimate(values, style) {
    if (style === 'mean') {
        return values.reduce((a, b) => a + b, 0) / values.length;
    }
    const sorted = [...values].sort((a, b) => a - b);
    if (style === 'median') {
        const mid = Math.floor(sorted.length / 2);
        return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) * 0.5;
    }
    return sorted.reduce((sum, v, i) => sum + v * weights(sorted.length)[i], 0);
}

function manualReference(
    data,
    length = 25,
    offset = 0,
    multiplier = 2.0,
    slopeStyle = 'smooth_median',
    residualStyle = 'smooth_median',
    deviationStyle = 'mad',
    madStyle = 'smooth_median',
    includePredictionInDeviation = false,
) {
    const n = data.length;
    const out = {
        value: new Float64Array(n),
        upper: new Float64Array(n),
        lower: new Float64Array(n),
        slope: new Float64Array(n),
        intercept: new Float64Array(n),
        deviation: new Float64Array(n),
    };
    for (const key of Object.keys(out)) out[key].fill(NaN);
    const warmup = length + offset - 1;

    for (let idx = warmup; idx < n; idx++) {
        const base = idx - offset;
        const start = base + 1 - length;
        const end = includePredictionInDeviation ? idx : base;
        let finite = true;
        for (let i = start; i <= end; i++) {
            if (!Number.isFinite(data[i])) {
                finite = false;
                break;
            }
        }
        if (!finite) continue;

        const slopes = [];
        for (let i = 0; i < length - 1; i++) {
            const valueI = data[base - i];
            for (let j = i + 1; j < length; j++) {
                slopes.push((data[base - j] - valueI) / (j - i));
            }
        }
        const beta1 = estimate(slopes, slopeStyle);
        const residuals = [];
        for (let j = 0; j < length; j++) residuals.push(data[base - j] - beta1 * j);
        const beta0 = estimate(residuals, residualStyle);
        const predicted = beta0 - beta1 * offset;

        let dev;
        if (deviationStyle === 'mad') {
            const errors = [];
            const pointStart = includePredictionInDeviation ? -offset : 0;
            for (let point = pointStart; point < length; point++) {
                const sourceIdx = idx - offset - point;
                const predictedPoint = beta0 + beta1 * point;
                errors.push(Math.abs(data[sourceIdx] - predictedPoint));
            }
            dev = estimate(errors, madStyle) * multiplier;
        } else {
            let sq = 0;
            let count = 0;
            const pointStart = includePredictionInDeviation ? -offset : 0;
            for (let point = pointStart; point < length; point++) {
                const sourceIdx = idx - offset - point;
                const predictedPoint = beta0 + beta1 * point;
                const diff = data[sourceIdx] - predictedPoint;
                sq += diff * diff;
                count += 1;
            }
            dev = Math.sqrt(sq / count) * multiplier;
        }

        out.value[idx] = predicted;
        out.upper[idx] = predicted + dev;
        out.lower[idx] = predicted - dev;
        out.slope[idx] = beta1;
        out.intercept[idx] = beta0;
        out.deviation[idx] = dev;
    }

    return out;
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('smooth_theil_sen_js matches manual reference', () => {
    const data = new Float64Array(testData.close.slice(0, 192));
    const result = wasm.smooth_theil_sen_js(
        data, 21, 2, 1.75, 'smooth_median', 'median', 'mad', 'smooth_median', true,
    );
    const expected = manualReference(
        data, 21, 2, 1.75, 'smooth_median', 'median', 'mad', 'smooth_median', true,
    );
    for (const key of ['value', 'upper', 'lower', 'slope', 'intercept', 'deviation']) {
        assertArrayClose(result[key], expected[key], 1e-10, `${key} mismatch`);
    }
});

test('smooth_theil_sen_batch_js first row matches single', () => {
    const data = new Float64Array(testData.close.slice(0, 144));
    const batch = wasm.smooth_theil_sen_batch_js(data, {
        length_range: [21, 23, 2],
        offset_range: [1, 1, 0],
        multiplier_range: [1.5, 1.5, 0.0],
        slope_style: 'mean',
        residual_style: 'smooth_median',
        deviation_style: 'rmsd',
        mad_style: 'median',
        include_prediction_in_deviation: false,
    });
    const single = wasm.smooth_theil_sen_js(
        data, 21, 1, 1.5, 'mean', 'smooth_median', 'rmsd', 'median', false,
    );
    assert.strictEqual(batch.rows, 2);
    assert.strictEqual(batch.cols, data.length);
    assertArrayClose(batch.lengths, [21, 23], 0, 'length grid mismatch');
    assertArrayClose(batch.offsets, [1, 1], 0, 'offset grid mismatch');
    assertArrayClose(batch.multipliers, [1.5, 1.5], 1e-12, 'multiplier grid mismatch');
    for (const key of ['value', 'upper', 'lower', 'slope', 'intercept', 'deviation']) {
        assertArrayClose(batch[key].slice(0, data.length), single[key], 1e-10, `${key} batch mismatch`);
    }
});

test('smooth_theil_sen_into pointer path matches safe API', () => {
    const data = new Float64Array(testData.close.slice(0, 160));
    const safe = wasm.smooth_theil_sen_js(
        data, 21, 2, 1.5, 'smooth_median', 'median', 'mad', 'smooth_median', true,
    );
    const dataPtr = wasm.allocate_f64_array(data.length);
    const valuePtr = wasm.smooth_theil_sen_alloc(data.length);
    const upperPtr = wasm.smooth_theil_sen_alloc(data.length);
    const lowerPtr = wasm.smooth_theil_sen_alloc(data.length);
    const slopePtr = wasm.smooth_theil_sen_alloc(data.length);
    const interceptPtr = wasm.smooth_theil_sen_alloc(data.length);
    const deviationPtr = wasm.smooth_theil_sen_alloc(data.length);
    try {
        wasm.write_f64_array(dataPtr, data);
        wasm.smooth_theil_sen_into(
            dataPtr, valuePtr, upperPtr, lowerPtr, slopePtr, interceptPtr, deviationPtr,
            data.length, 21, 2, 1.5, 'smooth_median', 'median', 'mad', 'smooth_median', true,
        );
        assertArrayClose(wasm.read_f64_array(valuePtr, data.length), safe.value, 1e-10, 'value pointer mismatch');
        assertArrayClose(wasm.read_f64_array(upperPtr, data.length), safe.upper, 1e-10, 'upper pointer mismatch');
        assertArrayClose(wasm.read_f64_array(lowerPtr, data.length), safe.lower, 1e-10, 'lower pointer mismatch');
        assertArrayClose(wasm.read_f64_array(slopePtr, data.length), safe.slope, 1e-10, 'slope pointer mismatch');
        assertArrayClose(wasm.read_f64_array(interceptPtr, data.length), safe.intercept, 1e-10, 'intercept pointer mismatch');
        assertArrayClose(wasm.read_f64_array(deviationPtr, data.length), safe.deviation, 1e-10, 'deviation pointer mismatch');
    } finally {
        wasm.deallocate_f64_array(dataPtr);
        wasm.smooth_theil_sen_free(valuePtr, data.length);
        wasm.smooth_theil_sen_free(upperPtr, data.length);
        wasm.smooth_theil_sen_free(lowerPtr, data.length);
        wasm.smooth_theil_sen_free(slopePtr, data.length);
        wasm.smooth_theil_sen_free(interceptPtr, data.length);
        wasm.smooth_theil_sen_free(deviationPtr, data.length);
    }
});

test('smooth_theil_sen_batch_into returns row count', () => {
    const data = new Float64Array(testData.close.slice(0, 128));
    const rows = 2;
    const dataPtr = wasm.allocate_f64_array(data.length);
    const valuePtr = wasm.smooth_theil_sen_alloc(rows * data.length);
    const upperPtr = wasm.smooth_theil_sen_alloc(rows * data.length);
    const lowerPtr = wasm.smooth_theil_sen_alloc(rows * data.length);
    const slopePtr = wasm.smooth_theil_sen_alloc(rows * data.length);
    const interceptPtr = wasm.smooth_theil_sen_alloc(rows * data.length);
    const deviationPtr = wasm.smooth_theil_sen_alloc(rows * data.length);
    try {
        wasm.write_f64_array(dataPtr, data);
        const returnedRows = wasm.smooth_theil_sen_batch_into(
            dataPtr, valuePtr, upperPtr, lowerPtr, slopePtr, interceptPtr, deviationPtr,
            data.length, 21, 23, 2, 1, 1, 0, 1.5, 1.5, 0.0, 'mean', 'smooth_median', 'rmsd', 'median', false,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.deallocate_f64_array(dataPtr);
        wasm.smooth_theil_sen_free(valuePtr, rows * data.length);
        wasm.smooth_theil_sen_free(upperPtr, rows * data.length);
        wasm.smooth_theil_sen_free(lowerPtr, rows * data.length);
        wasm.smooth_theil_sen_free(slopePtr, rows * data.length);
        wasm.smooth_theil_sen_free(interceptPtr, rows * data.length);
        wasm.smooth_theil_sen_free(deviationPtr, rows * data.length);
    }
});

test('smooth_theil_sen_js rejects invalid style', () => {
    const data = new Float64Array(testData.close.slice(0, 64));
    assert.throws(
        () => wasm.smooth_theil_sen_js(
            data, 21, 0, 2.0, 'bad_style', 'smooth_median', 'mad', 'smooth_median', false,
        ),
        /Invalid stat style/,
    );
});

File: tests/python/test_smooth_theil_sen.py
import math
import numpy as np
import pytest

from test_utils import load_test_data

try:
    import vector_ta as mp
except ImportError:
    try:
        import my_project as mp
    except ImportError:
        pytest.skip(
            "Python module not built. Run 'maturin develop --features python' first",
            allow_module_level=True,
        )


def _weights(size):
    center = (size - 1) * 0.5
    width = (size - 0.5) * (1024 ** (-0.5)) + 0.5
    values = np.array(
        [math.exp(-(((i - center) / width) ** 2) / 4.0) / math.sqrt(2.0 * math.pi) for i in range(size)],
        dtype=np.float64,
    )
    return values / values.sum()


def _estimate(values, style):
    data = np.asarray(values, dtype=np.float64)
    if style == "mean":
        return float(data.mean())
    data = np.sort(data)
    if style == "median":
        mid = data.shape[0] // 2
        if data.shape[0] % 2:
            return float(data[mid])
        return float((data[mid - 1] + data[mid]) * 0.5)
    return float(np.dot(data, _weights(data.shape[0])))


def manual_reference(
    data,
    length=25,
    offset=0,
    multiplier=2.0,
    slope_style="smooth_median",
    residual_style="smooth_median",
    deviation_style="mad",
    mad_style="smooth_median",
    include_prediction_in_deviation=False,
):
    data = np.asarray(data, dtype=np.float64)
    n = data.shape[0]
    out = {
        "value": np.full(n, np.nan, dtype=np.float64),
        "upper": np.full(n, np.nan, dtype=np.float64),
        "lower": np.full(n, np.nan, dtype=np.float64),
        "slope": np.full(n, np.nan, dtype=np.float64),
        "intercept": np.full(n, np.nan, dtype=np.float64),
        "deviation": np.full(n, np.nan, dtype=np.float64),
    }
    warmup = length + offset - 1

    for idx in range(warmup, n):
        base = idx - offset
        start = base + 1 - length
        end = idx if include_prediction_in_deviation else base
        segment = data[start : end + 1]
        if not np.all(np.isfinite(segment)):
            continue

        slopes = []
        for i in range(length - 1):
            value_i = data[base - i]
            for j in range(i + 1, length):
                slopes.append((data[base - j] - value_i) / (j - i))
        beta_1 = _estimate(slopes, slope_style)

        residuals = [data[base - j] - beta_1 * j for j in range(length)]
        beta_0 = _estimate(residuals, residual_style)
        predicted = beta_0 - beta_1 * offset

        if deviation_style == "mad":
            point_start = -offset if include_prediction_in_deviation else 0
            errors = []
            for point in range(point_start, length):
                source_idx = idx - offset - point
                predicted_point = beta_0 + beta_1 * point
                errors.append(abs(data[source_idx] - predicted_point))
            dev = _estimate(errors, mad_style) * multiplier
        else:
            point_start = -offset if include_prediction_in_deviation else 0
            sq = 0.0
            count = 0
            for point in range(point_start, length):
                source_idx = idx - offset - point
                predicted_point = beta_0 + beta_1 * point
                diff = data[source_idx] - predicted_point
                sq += diff * diff
                count += 1
            dev = math.sqrt(sq / count) * multiplier

        out["value"][idx] = predicted
        out["upper"][idx] = predicted + dev
        out["lower"][idx] = predicted - dev
        out["slope"][idx] = beta_1
        out["intercept"][idx] = beta_0
        out["deviation"][idx] = dev

    return out


class TestSmoothTheilSen:
    def test_manual_reference_matches_binding(self):
        data = np.asarray(load_test_data()["close"][:192], dtype=np.float64)
        expected = manual_reference(
            data,
            length=21,
            offset=2,
            multiplier=1.75,
            slope_style="smooth_median",
            residual_style="median",
            deviation_style="mad",
            mad_style="smooth_median",
            include_prediction_in_deviation=True,
        )
        result = mp.smooth_theil_sen(
            data,
            length=21,
            offset=2,
            multiplier=1.75,
            slope_style="smooth_median",
            residual_style="median",
            deviation_style="mad",
            mad_style="smooth_median",
            include_prediction_in_deviation=True,
        )
        for key in expected:
            np.testing.assert_allclose(result[key], expected[key], rtol=0.0, atol=1e-10, equal_nan=True)

    def test_stream_matches_batch(self):
        data = np.asarray(load_test_data()["close"][:160], dtype=np.float64)
        batch = mp.smooth_theil_sen(
            data,
            length=21,
            offset=2,
            multiplier=1.5,
            slope_style="smooth_median",
            residual_style="median",
            deviation_style="mad",
            mad_style="smooth_median",
            include_prediction_in_deviation=True,
        )
        stream = mp.SmoothTheilSenStream(
            21, 2, 1.5, "smooth_median", "median", "mad", "smooth_median", True
        )
        streamed = np.array([stream.update(float(v)) for v in data], dtype=np.float64)
        np.testing.assert_allclose(streamed[:, 0], batch["value"], rtol=0.0, atol=1e-10, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 1], batch["upper"], rtol=0.0, atol=1e-10, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 2], batch["lower"], rtol=0.0, atol=1e-10, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 3], batch["slope"], rtol=0.0, atol=1e-10, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 4], batch["intercept"], rtol=0.0, atol=1e-10, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 5], batch["deviation"], rtol=0.0, atol=1e-10, equal_nan=True)

    def test_batch_metadata_and_first_row_match_single(self):
        data = np.asarray(load_test_data()["close"][:144], dtype=np.float64)
        result = mp.smooth_theil_sen_batch(
            data,
            length_range=(21, 23, 2),
            offset_range=(1, 1, 0),
            multiplier_range=(1.5, 1.5, 0.0),
            slope_style="mean",
            residual_style="smooth_median",
            deviation_style="rmsd",
            mad_style="median",
            include_prediction_in_deviation=False,
        )
        single = mp.smooth_theil_sen(
            data,
            length=21,
            offset=1,
            multiplier=1.5,
            slope_style="mean",
            residual_style="smooth_median",
            deviation_style="rmsd",
            mad_style="median",
            include_prediction_in_deviation=False,
        )
        assert result["rows"] == 2
        assert result["cols"] == data.shape[0]
        np.testing.assert_allclose(result["lengths"], np.array([21, 23]), rtol=0.0, atol=0.0)
        np.testing.assert_allclose(result["offsets"], np.array([1, 1]), rtol=0.0, atol=0.0)
        np.testing.assert_allclose(result["multipliers"], np.array([1.5, 1.5]), rtol=0.0, atol=1e-12)
        for key in ["value", "upper", "lower", "slope", "intercept", "deviation"]:
            np.testing.assert_allclose(result[key][0], single[key], rtol=0.0, atol=1e-10, equal_nan=True)

    def test_invalid_style(self):
        data = np.asarray(load_test_data()["close"][:64], dtype=np.float64)
        with pytest.raises(ValueError, match="Invalid slope_style"):
            mp.smooth_theil_sen(data, slope_style="bad_style")
